aten/src/ATen/native/cuda/DilatedMaxPool2d.cu

```diff
@@ -283,44 +283,36 @@ __global__ void max_pool_backward_nhwc(const int nthreads, const scalar_t* top_d
   }
 }
 
-void max_pool2d_with_indices_out_cuda_template(
-           Tensor& output,
-           Tensor& indices,
-           const Tensor& input_,
-           IntArrayRef kernel_size,
-           IntArrayRef stride,
-           IntArrayRef padding,
-           IntArrayRef dilation,
-           bool ceil_mode)
-{
+} // namespace
+
+TORCH_IMPL_FUNC(max_pool2d_with_indices_out_cuda)
+(const Tensor& input_,
+IntArrayRef kernel_size,
+IntArrayRef stride,
+IntArrayRef padding,
+IntArrayRef dilation,
+bool ceil_mode,
+const Tensor& output,
+const Tensor& indices) {
```

Comment:
@ezyang Why the result must be defined as `const Tensor&`? I think `Tensor&` is more suitable. Because some internal function call(like `at::native::copy_`) will be `binding reference of type "at::Tensor &" to initializer of type "const at::Tensor"` when we use `const Tensor& result`.  
This will be a big problem when porting other kernel to structured kernel.

Reply:
This is trouble with incremental refactoring. If we could change everything, all at once, to `const Tensor&`, you wouldn't have this problem, but that's a ton of refactoring so we can only do things incrementally, which results in goofiness like this. The state approved answer is to `const_cast` when you hit a situation like this, and we'll fix them later.